botocore/useragent.py

```diff
@@ -0,0 +1,421 @@
+        self._session_user_agent_extra = session_user_agent_extra
+        return self
+
+    def with_client_config(self, client_config):
+        """
+        Create a copy with all original values and client-specific values.
+
+        :type client_config: botocore.config.Config
+        :param client_config: The client configuration object.
+        """
+        cp = copy(self)
+        cp._client_config = client_config
+        return cp
+
+    def to_string(self):
+        """
+        Build User-Agent header string from the object's properties.
+        """
+        if self._client_config and self._client_config.user_agent:
+            return self._build_legacy_ua_string()
+
+        components = [
+            *self._build_sdk_metadata(),
+            RawStringUserAgentComponent('ua/2.0'),
+            *self._build_os_metadata(),
+            *self._build_language_metadata(),
+            *self._build_execution_env_metadata(),
+            *self._build_feature_metadata(),
+            *self._build_config_metadata(),
+            *self._build_app_id(),
+            *self._build_extra(),
+        ]
+        return ' '.join([comp.to_string() for comp in components])
+
+    def _build_sdk_metadata(self):
+        """
+        Build the SDK name and version component of the User-Agent header.
+
+        For backwards-compatibility both session-level and client-level config
+        of custom tool names are honored. If this removes the Botocore
+        information from the start of the string, Botocore's name and version
+        are included as a separate field with "md" prefix.
+        """
+        if (
+            self._session_user_agent_name
+            and self._session_user_agent_version
+            and (
+                self._session_user_agent_name != _USERAGENT_SDK_NAME
+                or self._session_user_agent_version != botocore_version
+            )
+        ):
+            return [
+                UserAgentComponent(
+                    self._session_user_agent_name,
+                    self._session_user_agent_version,
+                ),
+                UserAgentComponent(
+                    'md', _USERAGENT_SDK_NAME, botocore_version
+                ),
+            ]
+
+        return [UserAgentComponent(_USERAGENT_SDK_NAME, botocore_version)]
+
+    def _build_os_metadata(self):
+        """
+        Build the OS/platform components of the User-Agent header string.
+
+        For recognized platform names that match or map to an entry in the list
+        of standardized OS names, a single component with prefix "os" is
+        returned. Otherwise, one component "os/other" is returned and a second
+        with prefix "md" and the raw platform name.
+
+        String representations of example return values:
+         * ``os/macos#10.13.6``
+         * ``os/linux``
+         * ``os/other``
+         * ``os/other md/foobar#1.2.3``
+         * ``os/linux md/arch#x86_64``
```

Comment:
We'd talked about this elsewhere, but this still strikes me as somewhat odd. Is `md/` that follows a more specific prefix always related or do you need to have implicit knowledge to associate entries?

Edit: Actually looking at the code below, how does the `arch` case work?

We either return one entry that's `os/{os_family}#{self._platform_version}` or `os/other md/{...}`. I don't see how we'd get linux + the md but maybe I'm missing it 😅

Reply:
> Edit: Actually looking at the code below, how does the arch case work?

Fixed in e19e237

> Is md/ that follows a more specific prefix always related or do you need to have implicit knowledge to associate entries?

The idea is that they are related. Or at least that's my understanding of how the spec came about. I agree that this example of `arch` following `os` seems like a tenuous relationship.